birdhouse/read-configs.include.sh

```diff
@@ -81,28 +81,50 @@ read_env_local() {
 
 }
 
+source_conf_files() {
+  dirs=$1
+  conf_locations=$2
+  for adir in ${dirs}; do
+      if echo "$ALL_CONF_DIRS" | grep -qE "^\s*$adir\s*$"; then
+          continue
+      fi
+      if [ -e "$adir" ]; then
+          ALL_CONF_DIRS="$ALL_CONF_DIRS
+            $adir
+          "
+      else
+          # Do not exit to not break unattended autodeploy since no human around to
+          # fix immediately.
+          # The new adir with typo will not be active but at least all the existing
+          # will still work.
+          echo "WARNING: '$adir' in $conf_locations does not exist" 1>&2
+      fi
+      COMPONENT_DEFAULT_ENV="$adir/default.env"
+      if [ -f "$COMPONENT_DEFAULT_ENV" ]; then
+          echo "reading '$COMPONENT_DEFAULT_ENV'"
+          . "$COMPONENT_DEFAULT_ENV"
+      fi
+  done
+}
 
 read_components_default_env() {
-    # EXTRA_CONF_DIRS normally set by env.local so should read_env_local() first.
+    # EXTRA_CONF_DIRS and DEFAULT_CONF_DIRS normally set by env.local so should read_env_local() first.
 
-    # EXTRA_CONF_DIRS relative paths are relative to COMPOSE_DIR.
+    # EXTRA_CONF_DIRS and DEFAULT_CONF_DIRS relative paths are relative to COMPOSE_DIR.
     if [ -d "$COMPOSE_DIR" ]; then
         cd "$COMPOSE_DIR"
     fi
-
-    for adir in ${EXTRA_CONF_DIRS}; do
-        if [ ! -e "$adir" ]; then
-            # Do not exit to not break unattended autodeploy since no human around to
-            # fix immediately.
-            # The new adir with typo will not be active but at least all the existing
-            # will still work.
-            echo "WARNING: '$adir' in EXTRA_CONF_DIRS does not exist" 1>&2
-        fi
-        COMPONENT_DEFAULT_ENV="$adir/default.env"
-        if [ -f "$COMPONENT_DEFAULT_ENV" ]; then
-            echo "reading '$COMPONENT_DEFAULT_ENV'"
-            . "$COMPONENT_DEFAULT_ENV"
-        fi
+    requested_conf_dirs="
+      $DEFAULT_CONF_DIRS
+      $EXTRA_CONF_DIRS
+    "
+    ALL_CONF_DIRS=''
+    COMPONENT_DEPENDENCIES=''
+    current_dependencies=$COMPONENT_DEPENDENCIES
+    source_conf_files "$requested_conf_dirs" 'EXTRA_CONF_DIRS or DEFAULT_CONF_DIRS'
+    while [ "$current_dependencies" != "$COMPONENT_DEPENDENCIES" ]; do
```

Comment:
Does this loop ever trigger?  Initially `COMPONENT_DEPENDENCIES=''` and `current_dependencies=$COMPONENT_DEPENDENCIES` so the condtion `[ "$current_dependencies" != "$COMPONENT_DEPENDENCIES" ]` would be `false` so this loop would never fire.

But it's 22h30, my brain is shutting down, I might be wrong.

Reply:
`source_conf_files` can update `COMPONENT_DEPENDENCIES` in the line directly before the while loop (line 124). I'll add a clarifying comment.

Reply:
You're right, it was late that evening :D

Reply:
A comment above `source_conf_files ` call such as `# update COMPONENT_DEPENDENCIES with each component's sourced default.env` will greatly help to read this part, as `COMPONENT_DEPENDENCIES` doesn't appear directly in the `source_conf_files` function.